packages/library-base/src/commonMain/kotlin/io/realm/internal/query/ScalarQuery.kt

```diff
@@ -0,0 +1,144 @@
+ * limitations under the License.
+ */
+
+package io.realm.internal.query
+
+import io.realm.RealmInstant
+import io.realm.RealmObject
+import io.realm.RealmResults
+import io.realm.internal.Mediator
+import io.realm.internal.RealmReference
+import io.realm.internal.RealmResultsImpl
+import io.realm.internal.Thawable
+import io.realm.internal.genericRealmCoreExceptionHandler
+import io.realm.internal.interop.NativePointer
+import io.realm.internal.interop.PropertyKey
+import io.realm.internal.interop.RealmCoreException
+import io.realm.internal.interop.RealmCoreLogicException
+import io.realm.internal.interop.RealmInterop
+import io.realm.internal.interop.Timestamp
+import io.realm.query.RealmScalarQuery
+import kotlinx.coroutines.flow.Flow
+import kotlin.reflect.KClass
+
+internal enum class AggregatorQueryType {
+    MIN, MAX, SUM
+}
+
+/**
+ * Shared logic for scalar queries.
+ *
+ * Observe that this class needs the [E] representing a [RealmObject] to avoid having to split
+ * [RealmResults] in object and scalar implementations and to be able to observe changes to the
+ * scalar values for the query - more concretely to allow returning a [RealmResultsImpl] object by
+ * [thaw], which in turn comes from processing said results with [queryMapper].
+ */
+internal abstract class BaseScalarQuery<E : RealmObject, T : Any> constructor(
+    protected val realmReference: RealmReference,
+    protected val queryPointer: NativePointer,
+    protected val mediator: Mediator
+) : RealmScalarQuery<T>, Thawable<RealmResultsImpl<E>> {
+
+    abstract fun Flow<RealmResultsImpl<E>?>.queryMapper(): Flow<T?>
+    abstract fun discardRepeated(latestValue: T?): Boolean
+
+    override fun thaw(liveRealm: RealmReference): RealmResultsImpl<E> = TODO()
+
+    override fun asFlow(): Flow<T?> = TODO()
+
+    protected fun getPropertyKey(clazz: KClass<*>, property: String): PropertyKey =
+        RealmInterop.realm_get_col_key(realmReference.dbPointer, clazz.simpleName!!, property)
+}
+
+internal class CountQuery<E : RealmObject> constructor(
+    realmReference: RealmReference,
+    queryPointer: NativePointer,
+    mediator: Mediator
+) : BaseScalarQuery<E, Long>(realmReference, queryPointer, mediator) {
+
+    override fun find(): Long = RealmInterop.realm_query_count(queryPointer)
+
+    override fun Flow<RealmResultsImpl<E>?>.queryMapper(): Flow<Long?> = TODO()
+
+    override fun discardRepeated(latestValue: Long?): Boolean = TODO()
+}
+
+@Suppress("LongParameterList")
+internal class AggregatorQuery<E : RealmObject, T : Any> constructor(
+    realmReference: RealmReference,
+    queryPointer: NativePointer,
+    mediator: Mediator,
+    private val clazz: KClass<*>,
+    private val property: String,
+    private val type: KClass<T>,
+    private val queryType: AggregatorQueryType
+) : BaseScalarQuery<E, T>(realmReference, queryPointer, mediator) {
+
+    override fun find(): T? = findInternal(queryPointer)
```

Comment:
Any reason for not just inlining `findInternal` here. Doesn't seem to be used elsewhere.

Reply:
Yes, I noticed that later in the process. It should be there in https://github.com/realm/realm-kotlin/pull/609